### DISCUSSION TITLE
Should we require testing on Firefox?

### DISCUSSION THREAD

**Discussion by DNin01** [ScratchAddons/ScratchAddons#5810] [Discussion]
This is a conversation about whether to adapt a new requirement for PRs related to testing code on Firefox.

### Let's set up a requirement

There have been several past issues with Scratch Addons that only happened on Firefox, as you can see [here](https://github.com/ScratchAddons/ScratchAddons/issues?q=label%3A%22platform%3A+firefox%22+). To help reduce the number of Firefox-exclusive issues, it's worth discussing testing on Firefox. I have a few ideas:
<br>

#### 👍 _Require_ testing on both Chrome and Firefox to approve code changes (stricter)
**Reviewers of any PR that submits code changes would need to test the changes on both a Chromium-based browser and Firefox to give a valid approval.**

This immediately would greatly reduce the number of Firefox-specific issues, but does every PR really need to be checked that thoroughly? In my opinion, one test for each browser should be enough.
<br>

#### 🚀 _Require_ testing on both Chrome and Firefox to merge code changes (strict)
**Any PR that submits code changes must have been tested on both a Chromium-based browser and Firefox before merged.**

This is a decent balance and ensures changes work well on Firefox without making the review process as strict. This may be unnecessary for some code changes, but we could handle it.
<br>

#### 🥗 _Require_ testing on both Chrome and Firefox to merge complex code changes (moderate)
**Any PR that 1) submits code changes to one or more core, critical or important parts of SA (like the full screen settings, a frequently used addon, a library or an API), 2) submits large code changes or a large number of code changes or 3) is likely to not translate well into other browsers must have been tested on both a Chromium-based browser and Firefox before merged.**

If doable, having at least this makes the most sense for our project at this time, unless we have a reason to be more thorough about testing.
<br>

#### 👋 _Recommend_ testing on Firefox (current)
**Testing on a Chromium-based browser would be required, but testing on Firefox would be ideal but not necessary**

This could result in issues that slip into future releases, which, you know, no one likes.
<br>

Let me know what you think would be the most helpful, or if you have any other ideas.

### Task list

We should also add a task list to the PR template that would encourage the PR author to test their code changes on both browsers. The description would be adjusted to fit the requirement we choose.

```
### Tests

Please check off the following:
- [ ] Tested on a Chromium-based browser
- [ ] Tested on Firefox
```

Reviewers could also interact with that task list - if the PR author did not test on one of the browsers, a reviewer could do it for them and check off the task for the test afterward.

If we want to get really thorough, we might be able to use a status check to block merging until the tasks are complete.

### Related

- https://github.com/ScratchAddons/website-v2/pull/219
- #4956

**Comment by Samq64** [comment]
> :rocket: Require testing on both Chrome and Firefox to merge code changes (strict)

I think that's the best option.

Also, I think the browser checklist could be added to the PR template now regardless of what we choose.

**Reply by BroJac5246** [reply]
I agree. Not everyone has access to Firefox (for example, chromeOS users who don't want to set up Linux, etc.), but we need to ensure compatibility with Firefox, which should be accomplished by requiring a test on both it and a Chromium browser. Having the checklist in the PR description (like what Scratch uses on GitHub) and updating it after reviews would be a great way to track this.

**Comment by Hans5958** [comment]
If we can utilize something like BrowserStack then this would be easier.
